### PR TITLE
Use leveldb to store ValidatorEnodes information 

### LINKED ISSUE #193: Validators SBAT persistently cache others validators' nodes
### Expected Behavior

Depends on #191 

Store cache of other validators' enodes on disk so we can pre-populate cache on startup and quickly rejoin consensus protocol.

### Current Behavior

With #191, we would forget all details of other validators we'd seen if we restart.

### PR DESCRIPTION
**Pull request by mcortesi**
### Description

Summary of changes:

- rename: val_enode_table => val_enode_db
- new istanbul.Config field: ValidatorEnodeDBPath
- Use leveldb inside val_enode_db
- Remove locks inside val_enode_db (leveldb can be used concurrently)
- Handle/propagate leveldb errors
- RLP encode/decode for table entries.

### Tested

unit test

### Related issues

- Fixes #193

### Backwards compatibility

Yes

### CODE REVIEW

File: consensus/istanbul/backend/announce.go [hunk 1]
```diff
@@ -189,12 +192,14 @@ func (sb *Backend) generateIstAnnounce() ([]byte, error) {
 			} else {
 				encryptedEndpoints = append(encryptedEndpoints, [][]byte{addr.Bytes(), encryptedEndpoint})
 			}
+		} else if err != leveldb.ErrNotFound {
+			log.Warn("Unable to read valEnodeTable", "err", err, "addr", addr)
```
Comment (asaj) [inline consensus/istanbul/backend/announce.go]: Should this be `log.Error`?

File: consensus/istanbul/backend/backend.go [hunk 2]
```diff
@@ -542,17 +549,23 @@ func (sb *Backend) RefreshValPeers(valset istanbul.ValidatorSet) {
 	} else {
 		// Add all of the valSet entries as validator peers
 		for _, val := range valset.List() {
-			if enodeURL, ok := sb.valEnodeTable.GetEnodeURLFromAddress(val.Address()); ok {
+			enodeURL, err := sb.valEnodeTable.GetEnodeURLFromAddress(val.Address())
+			if err == nil {
 				sb.AddValidatorPeer(enodeURL)
+			} else if err != leveldb.ErrNotFound {
+				sb.logger.Warn("Error reading valEnodeTable: GetEnodeURLFromAddress", "err", err)
```
Comment (asaj) [inline consensus/istanbul/backend/backend.go]: Ditto?

File: consensus/istanbul/backend/backend.go [hunk 3]
```diff
@@ -542,17 +549,23 @@ func (sb *Backend) RefreshValPeers(valset istanbul.ValidatorSet) {
...
 		// Remove the peers that are not in the valset
 		for _, peerEnodeURL := range currentValPeers {
-			if peerAddress, ok := sb.valEnodeTable.GetAddressFromEnodeURL(peerEnodeURL); ok {
+			peerAddress, err := sb.valEnodeTable.GetAddressFromEnodeURL(peerEnodeURL)
+			if err == nil {
 				if _, src := valset.GetByAddress(peerAddress); src == nil {
 					sb.RemoveValidatorPeer(peerEnodeURL)
 				}
+			} else if err != leveldb.ErrNotFound {
+				sb.logger.Warn("Error reading valEnodeTable: GetEnodeURLFromAddress", "err", err)
```
Comment (asaj) [inline consensus/istanbul/backend/backend.go]: Ditto?

File: consensus/istanbul/backend/internal/enodes/val_enode_db.go [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+	})
+
+	if err != nil {
+		log.Error("ValidatorEnodeDB.String error", "err", err)
+	}
+
+	return b.String()
+}
+
+// GetEnodeURLFromAddress will return the enodeURL for a address if it's kwown
```
Comment (asaj) [inline consensus/istanbul/backend/internal/enodes/val_enode_db.go]: "known"

File: consensus/istanbul/backend/internal/enodes/val_enode_db.go [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+func (vet *ValidatorEnodeDB) GetEnodeURLFromAddress(address common.Address) (string, error) {
+	entry, err := vet.getAddressEntry(address)
+
+	if err != nil {
+		return "", err
+	}
+	return entry.enodeURL, nil
+}
+
+// GetAddressFromEnodeURL will return the address for a enodeURL if it's kwown
```
Comment (asaj) [inline consensus/istanbul/backend/internal/enodes/val_enode_db.go]: ditto

File: consensus/istanbul/backend/internal/enodes/val_enode_db.go (hunk 4 above)
Comment (asaj) [inline consensus/istanbul/backend/internal/enodes/val_enode_db.go]: "an address"

File: consensus/istanbul/backend/internal/enodes/val_enode_db.go (hunk 5 above)
Comment (asaj) [inline consensus/istanbul/backend/internal/enodes/val_enode_db.go]: "an enodeURL"

File: consensus/istanbul/config.go [hunk 6]
```diff
@@ -24,15 +24,17 @@ const (
...
 type Config struct {
-	RequestTimeout uint64         `toml:",omitempty"` // The timeout for each Istanbul round in milliseconds.
-	BlockPeriod    uint64         `toml:",omitempty"` // Default minimum difference between two consecutive block's timestamps in second
-	ProposerPolicy ProposerPolicy `toml:",omitempty"` // The policy for proposer selection
-	Epoch          uint64         `toml:",omitempty"` // The number of blocks after which to checkpoint and reset the pending votes
+	RequestTimeout       uint64         `toml:",omitempty"` // The timeout for each Istanbul round in milliseconds.
+	BlockPeriod          uint64         `toml:",omitempty"` // Default minimum difference between two consecutive block's timestamps in second
+	ProposerPolicy       ProposerPolicy `toml:",omitempty"` // The policy for proposer selection
+	Epoch                uint64         `toml:",omitempty"` // The number of blocks after which to checkpoint and reset the pending votes
+	ValidatorEnodeDBPath string         `toml:",omitempty"` // The location for the validator enode's DB
```
Comment (asaj) [inline consensus/istanbul/config.go]: nit: "validator enodes"

File: consensus/istanbul/backend/internal/enodes/val_enode_db.go [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+	rawEntry, err := vet.db.Get(enodeURLKey(enodeURL), nil)
+	if err != nil {
+		return common.ZeroAddress, err
+	}
+	return common.BytesToAddress(rawEntry), nil
+}
+
+// Upsert will update or insert a validator enode entry; given that the existing entry
+// is older (determined by view parameter) that the new one
+func (vet *ValidatorEnodeDB) Upsert(remoteAddress common.Address, enodeURL string, view *istanbul.View) (string, error) {
```
Comment (kevjue) [inline consensus/istanbul/backend/internal/enodes/val_enode_db.go]: Should this deal with the case if the entry is not changed?

Comment (mcortesi) [reply consensus/istanbul/backend/internal/enodes/val_enode_db.go]: done!

**Comment by mcortesi** [comment]
> I only have one comment. Please take a look.
> 
> Also, when merging the Proxy PR, there is also the case of this change: https://github.com/celo-org/celo-blockchain/pull/515/files#diff-c88ce221086edfa0526423184f0035b1R199
> 
> When the proxy processes the val_enode_share message, it will currently grab the val_enode_share lock, call upsert individually on the shared vet, then unlock. We'll have to rework the `upsert` function a bit when doing the merge.

we will need a "batch update" function within the DB on the proxy PR

**Pull request merged**